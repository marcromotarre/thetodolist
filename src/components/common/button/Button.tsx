import { ReactElement, cloneElement, useState } from "react";
import Typography from "../typography/Typography";
import BUTTON_DATA from "./ButtonData";

const Button: React.FC<ComponentProps> = (props) => {
  const [isHovering, setIsHovering] = useState<boolean>(false);

  const buttonData = BUTTON_DATA({ ...props, isHovering });

  const handleMouseEnter = () => {
    setIsHovering(true);
  };

  const handleMouseLeave = () => {
    setIsHovering(false);
  };

  return (
    <button
      disabled={props.disabled}
      data-testid="button"
      onClick={props.onClick}
      style={buttonData.button.style}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      className={buttonData.button.className}
    >
      <div
        className={buttonData.buttonContent.className}
        data-testid="button-content"
      >
        {props.startIcon &&
          cloneElement(props.startIcon, {
            ...buttonData.buttonStartIcon.iconProps,
          })}
        {props.children && (
          <Typography
            className={buttonData.buttonText.className}
            variant="button"
            style={buttonData.buttonText.style}
          >
            {props.children}
          </Typography>
        )}
        {props.endIcon &&
          cloneElement(props.endIcon, {
            ...buttonData.buttonEndIcon.iconProps,
          })}
      </div>
    </button>
  );
};

export default Button;

export type ComponentProps = {
  children?: string;
  variant?: "outlined" | "text" | "contained";
  shape?: "default" | "rect" | "rounded";
  startIcon?: ReactElement;
  endIcon?: ReactElement;
  color?: "error" | "default" | "success" | string;
  secondaryColor?: string;
  size?: "small" | "medium" | "large";
  onClick?: React.MouseEventHandler;
  disabled?: boolean;
};

Button.defaultProps = {
  variant: "contained",
  color: "default",
  secondaryColor: "white",
  size: "medium",
  shape: "default",
  disabled: false,
};
