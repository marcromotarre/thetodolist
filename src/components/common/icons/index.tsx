// lib icons
export { default as SuccessCircledIcon } from "./lib-icons/SuccessCircledIcon";
export { default as WarningIcon } from "./lib-icons/WarningIcon";
export { default as InfoCircledIcon } from "./lib-icons/InfoCircledIcon";
export { default as ErrorCircledIcon } from "./lib-icons/ErrorCircledIcon";
export { default as PlusIcon } from "./lib-icons/PlusIcon";
export { default as CalendarIcon } from "./lib-icons/CalendarIcon";

// app icons
export { default as ComputerIcon } from "./app-icons/ComputerIcon";
export { default as PhoneIcon } from "./app-icons/PhoneIcon";
