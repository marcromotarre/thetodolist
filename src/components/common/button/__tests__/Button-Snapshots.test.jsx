import { expect, test, vi } from "vitest";
import { render } from "@testing-library/react";
import Button from "../Button";
import { RectangleIcon } from "../../icons";

test("renders correct text button", async () => {
  const { asFragment } = render(<Button>TextButton</Button>);
  expect(asFragment()).toMatchSnapshot();
});

test("renders correct icon button", async () => {
  const { asFragment } = render(<Button startIcon={<RectangleIcon />} />);
  expect(asFragment()).toMatchSnapshot();
});

test("renders correct startIcon text button", async () => {
  const { asFragment } = render(
    <Button startIcon={<RectangleIcon />}>TextButton</Button>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders correct endIcon text button", async () => {
  const { asFragment } = render(
    <Button endIcon={<RectangleIcon />}>TextButton</Button>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders correct small button", async () => {
  const { asFragment } = render(
    <Button size="small" startIcon={<RectangleIcon />}>
      Small Button
    </Button>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders correct medium button", async () => {
  const { asFragment } = render(
    <Button size="medium" startIcon={<RectangleIcon />}>
      Medium Button
    </Button>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders correct large button", async () => {
  const { asFragment } = render(
    <Button size="large" startIcon={<RectangleIcon />}>
      Large Button
    </Button>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders correct text variant button", async () => {
  const { asFragment } = render(
    <Button variant="text" startIcon={<RectangleIcon />}>
      Large Button
    </Button>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders correct text contained button", async () => {
  const { asFragment } = render(
    <Button variant="contained" startIcon={<RectangleIcon />}>
      Contained Button
    </Button>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders correct text outlined button", async () => {
  const { asFragment } = render(
    <Button variant="outlined" startIcon={<RectangleIcon />}>
      Contained Button
    </Button>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders correct color button", async () => {
  const { asFragment } = render(
    <Button color="red" secondaryColor="black" startIcon={<RectangleIcon />}>
      Contained Button
    </Button>
  );
  expect(asFragment()).toMatchSnapshot();
});
