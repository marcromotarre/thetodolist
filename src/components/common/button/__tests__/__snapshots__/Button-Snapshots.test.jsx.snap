// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correct color button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-3 rounded-md"
    data-testid="button"
    style="background-color: red;"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <svg
        class=""
        data-testid="rectangle-icon"
        fill="none"
        height="20"
        viewBox="0 0 72 72"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="black"
          height="72"
          width="72"
        />
      </svg>
      <span
        class="ml-2 mr-2 sm:text-xs md:text-sm lg:text-sm uppercase font-medium "
        data-testid="typography-span"
        style="font-size: 0.875rem;"
      >
        Contained Button
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders correct endIcon text button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-3 rounded-md"
    data-testid="button"
    style="background-color: rgb(96 165 250);"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <span
        class="ml-2 mr-2 sm:text-xs md:text-sm lg:text-sm uppercase font-medium "
        data-testid="typography-span"
        style="color: white; font-size: 0.875rem;"
      >
        TextButton
      </span>
      <svg
        class=""
        data-testid="rectangle-icon"
        fill="none"
        height="20"
        viewBox="0 0 72 72"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="white"
          height="72"
          width="72"
        />
      </svg>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders correct icon button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-3 rounded-md"
    data-testid="button"
    style="background-color: rgb(96 165 250);"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <svg
        class=""
        data-testid="rectangle-icon"
        fill="none"
        height="20"
        viewBox="0 0 72 72"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="white"
          height="72"
          width="72"
        />
      </svg>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders correct large button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-4 rounded-md"
    data-testid="button"
    style="background-color: rgb(96 165 250);"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <svg
        class=""
        data-testid="rectangle-icon"
        fill="none"
        height="25"
        viewBox="0 0 72 72"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="white"
          height="72"
          width="72"
        />
      </svg>
      <span
        class="ml-2 mr-2 sm:text-xs md:text-sm lg:text-sm uppercase font-medium "
        data-testid="typography-span"
        style="color: white; font-size: 1.25rem;"
      >
        Large Button
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders correct medium button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-3 rounded-md"
    data-testid="button"
    style="background-color: rgb(96 165 250);"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <svg
        class=""
        data-testid="rectangle-icon"
        fill="none"
        height="20"
        viewBox="0 0 72 72"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="white"
          height="72"
          width="72"
        />
      </svg>
      <span
        class="ml-2 mr-2 sm:text-xs md:text-sm lg:text-sm uppercase font-medium "
        data-testid="typography-span"
        style="color: white; font-size: 0.875rem;"
      >
        Medium Button
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders correct small button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-2 rounded-md"
    data-testid="button"
    style="background-color: rgb(96 165 250);"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <svg
        class=""
        data-testid="rectangle-icon"
        fill="none"
        height="15"
        viewBox="0 0 72 72"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="white"
          height="72"
          width="72"
        />
      </svg>
      <span
        class="ml-2 mr-2 sm:text-xs md:text-sm lg:text-sm uppercase font-medium "
        data-testid="typography-span"
        style="color: white; font-size: 0.7rem;"
      >
        Small Button
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders correct startIcon text button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-3 rounded-md"
    data-testid="button"
    style="background-color: rgb(96 165 250);"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <svg
        class=""
        data-testid="rectangle-icon"
        fill="none"
        height="20"
        viewBox="0 0 72 72"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="white"
          height="72"
          width="72"
        />
      </svg>
      <span
        class="ml-2 mr-2 sm:text-xs md:text-sm lg:text-sm uppercase font-medium "
        data-testid="typography-span"
        style="color: white; font-size: 0.875rem;"
      >
        TextButton
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders correct text button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-3 rounded-md"
    data-testid="button"
    style="background-color: rgb(96 165 250);"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <span
        class="ml-2 mr-2 sm:text-xs md:text-sm lg:text-sm uppercase font-medium "
        data-testid="typography-span"
        style="color: white; font-size: 0.875rem;"
      >
        TextButton
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders correct text contained button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-3 rounded-md"
    data-testid="button"
    style="background-color: rgb(96 165 250);"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <svg
        class=""
        data-testid="rectangle-icon"
        fill="none"
        height="20"
        viewBox="0 0 72 72"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="white"
          height="72"
          width="72"
        />
      </svg>
      <span
        class="ml-2 mr-2 sm:text-xs md:text-sm lg:text-sm uppercase font-medium "
        data-testid="typography-span"
        style="color: white; font-size: 0.875rem;"
      >
        Contained Button
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders correct text outlined button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-3 rounded-md"
    data-testid="button"
    style="background-color: transparent;"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <svg
        class=""
        data-testid="rectangle-icon"
        fill="none"
        height="20"
        viewBox="0 0 72 72"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="rgb(96 165 250)"
          height="72"
          width="72"
        />
      </svg>
      <span
        class="ml-2 mr-2 sm:text-xs md:text-sm lg:text-sm uppercase font-medium "
        data-testid="typography-span"
        style="color: rgb(96 165 250); font-size: 0.875rem;"
      >
        Contained Button
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`renders correct text variant button 1`] = `
<DocumentFragment>
  <button
    class="flex h-fit w-max items-center justify-center bg-transparent hover:bg-white hover:brightness-95 p-3 rounded-md"
    data-testid="button"
    style="background-color: transparent;"
  >
    <div
      class="flex items-center justify-center"
      data-testid="button-content"
    >
      <svg
        class=""
        data-testid="rectangle-icon"
        fill="none"
        height="20"
        viewBox="0 0 72 72"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="rgb(96 165 250)"
          height="72"
          width="72"
        />
      </svg>
      <span
        class="ml-2 mr-2 sm:text-xs md:text-sm lg:text-sm uppercase font-medium "
        data-testid="typography-span"
        style="color: rgb(96 165 250); font-size: 0.875rem;"
      >
        Large Button
      </span>
    </div>
  </button>
</DocumentFragment>
`;
